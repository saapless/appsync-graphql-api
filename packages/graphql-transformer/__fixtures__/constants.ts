import { DataSourceManagerConfig } from "../src/context/DataSourceManager";

export const TEST_DS_CONFIG = {
  primaryDataSourceName: "TestDataSource",
  dataSources: {
    TestDataSource: {
      type: "DYNAMO_DB",
      config: {
        tableName: "TestTable",
      },
    },
    NoneDataSource: {
      type: "NONE",
    },
  },
} satisfies DataSourceManagerConfig;
