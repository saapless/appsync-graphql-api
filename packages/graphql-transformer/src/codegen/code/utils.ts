/* eslint-disable security/detect-object-injection */

import { PropertyValue, tc } from "./ast";

export function formatValue(value: unknown): PropertyValue {
  if (typeof value === "string") {
    return tc.str(value);
  } else if (typeof value === "number") {
    return tc.num(value);
  } else if (typeof value === "boolean") {
    return tc.bool(value);
  } else if (Array.isArray(value)) {
    return tc.arr(...value.map((v) => formatValue(v)));
  } else if (value === null) {
    return tc.null();
  } else if (typeof value === "object") {
    return tc.obj(
      Object.entries(value).reduce(
        (agg, [n, v]) => {
          agg[n] = formatValue(v);
          return agg;
        },
        {} as Record<string, PropertyValue>
      )
    );
  } else {
    throw new Error("Invalid value");
  }
}
