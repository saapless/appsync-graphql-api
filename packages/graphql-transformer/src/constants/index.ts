export const ConnectionDirective = Object.freeze({
  HAS_ONE: "hasOne",
  HAS_MANY: "hasMany",
});

export const ModelDirective = Object.freeze({
  MODEL: "model",
});

export const UtilityDirective = Object.freeze({
  SERVER_ONLY: "serverOnly",
  CLIENT_ONLY: "clientOnly",
  READ_ONLY: "readOnly",
  WRITE_ONLY: "writeOnly",
  FILTER_ONLY: "filterOnly",
  INTERNAL: "internal",
});

export const ScalarType = Object.freeze({
  STRING: "String",
  INT: "Int",
  FLOAT: "Float",
  BOOLEAN: "Boolean",
  ID: "ID",
  LONG: "Long",
});

export const RelationType = Object.freeze({
  ONE_TO_ONE: "oneToOne",
  ONE_TO_MANY: "oneToMany",
  MANY_TO_MANY: "manyToMany",
});
