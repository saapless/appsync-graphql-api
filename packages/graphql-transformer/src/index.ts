export * from "./transformer";
export { definitionFromSource } from "./utils";
