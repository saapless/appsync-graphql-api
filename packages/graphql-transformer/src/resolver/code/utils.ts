export function wrap(start: string, value: string, end = "") {
  return `${start}${value}${end}`;
}

export function indent(value: string) {
  return value
    .split("\n")
    .map((line) => `  ${line}`)
    .join("\n");
}

export function join<T extends (string | string[])[]>(separator = "", ...statements: T) {
  return statements.flat().filter(Boolean).join(separator);
}

export function block<T extends string[]>(...value: T) {
  return wrap("{\n", indent(join("\n", value)).trimEnd(), "\n}\n");
}

export function expression(value: string) {
  return wrap("(", value, ")");
}

export function statement(value: string) {
  return wrap("", value, `${value.endsWith(";") ? "" : ";"}\n`);
}

export function returnStatement(value: string) {
  return wrap("return ", statement(value));
}
