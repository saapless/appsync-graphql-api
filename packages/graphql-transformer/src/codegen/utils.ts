/* eslint-disable security/detect-object-injection */

import { KeyValue } from "../utils/types";
import { PropertyValue, tc } from "./code";

export function formatValue(value: unknown): PropertyValue {
  if (typeof value === "string") {
    return tc.str(value);
  } else if (typeof value === "number") {
    return tc.num(value);
  } else if (typeof value === "boolean") {
    return tc.bool(value);
  } else if (Array.isArray(value)) {
    return tc.arr(...value.map((v) => formatValue(v)));
  } else if (value === null) {
    return tc.null();
  } else if (typeof value === "object") {
    return tc.obj(
      Object.entries(value).reduce(
        (agg, [n, v]) => {
          agg[n] = formatValue(v);
          return agg;
        },
        {} as Record<string, PropertyValue>
      )
    );
  } else {
    throw new Error("Invalid value");
  }
}

export function keyValue<T extends string | number>(obj: KeyValue<T>) {
  if (obj.ref) return tc.chain("ctx", obj.ref);
  if (obj.eq) return typeof obj.eq === "string" ? tc.str(obj.eq) : tc.num(obj.eq);
  throw new Error("Invalid key value");
}
