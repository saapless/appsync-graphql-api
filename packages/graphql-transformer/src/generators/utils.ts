/* eslint-disable security/detect-object-injection */

import { FieldLoaderDescriptor, Key, KeyValue, LoaderDescriptor } from "../utils/types";
import { PropertyValue, tc } from "../codegen";

export function formatValue(value: unknown): PropertyValue {
  if (typeof value === "string") {
    return tc.str(value);
  } else if (typeof value === "number") {
    return tc.num(value);
  } else if (typeof value === "boolean") {
    return tc.bool(value);
  } else if (Array.isArray(value)) {
    return tc.arr(...value.map((v) => formatValue(v)));
  } else if (value === null) {
    return tc.null();
  } else if (typeof value === "object") {
    return tc.obj(
      Object.entries(value).reduce(
        (agg, [n, v]) => {
          agg[n] = formatValue(v);
          return agg;
        },
        {} as Record<string, PropertyValue>
      )
    );
  } else {
    throw new Error("Invalid value");
  }
}

export function keyValue<T extends string | number>(obj: KeyValue<T>) {
  if (obj.ref) {
    // `ctx.source` is typed as optional in Context

    if (obj.ref.startsWith("source")) {
      obj.ref = obj.ref.replace("source", "source?");
    }

    return tc.chain("ctx", obj.ref);
  }
  if (obj.eq) return typeof obj.eq === "string" ? tc.str(obj.eq) : tc.num(obj.eq);
  throw new Error("Invalid key value");
}

export function isFieldLoader(descriptor: LoaderDescriptor): descriptor is FieldLoaderDescriptor {
  return Object.hasOwn(descriptor, "fieldName") && Object.hasOwn(descriptor, "typeName");
}

export function parseKey(key: Key) {
  return tc.obj(
    Object.entries(key).reduce(
      (agg, [name, op]) => {
        if (op.ref || op.eq) {
          Object.assign(agg, { [name]: keyValue(op) });
        } else if (op.le) {
          Object.assign(agg, { [name]: tc.obj({ le: keyValue(op.le) }) });
        } else if (op.lt) {
          Object.assign(agg, { [name]: tc.obj({ lt: keyValue(op.lt) }) });
        } else if (op.ge) {
          Object.assign(agg, { [name]: tc.obj({ ge: keyValue(op.ge) }) });
        } else if (op.gt) {
          Object.assign(agg, { [name]: tc.obj({ gt: keyValue(op.gt) }) });
        } else if (op.between) {
          Object.assign(agg, {
            [name]: tc.obj({ between: tc.arr(...op.between.map((i) => keyValue(i))) }),
          });
        } else if (op.beginsWith) {
          Object.assign(agg, { [name]: tc.obj({ beginsWith: keyValue(op.beginsWith) }) });
        }
        return agg;
      },
      {} as Record<string, PropertyValue>
    )
  );
}
