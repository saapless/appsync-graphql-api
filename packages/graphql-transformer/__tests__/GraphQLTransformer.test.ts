import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";
import { createTransformer, GraphQLTransformer } from "../src/transformer";
import { FieldNode, NamedTypeNode, ObjectNode } from "../src/definition";
import { SchemaValidationError } from "../src/utils/errors";
import { TEST_DS_CONFIG } from "../__fixtures__/constants";

const schema = /* GraphQL */ `
  enum UserStatus {
    ACTIVE
    DISABLED
    SUSPENDED
  }

  type User @model {
    id: ID!

    firstName: String
    lastName: String
    email: AWSEmail
    picture: AWSURL
    status: UserStatus @readOnly
  }

  type Task @model(operations: [upsert, delete]) {
    id: ID!

    title: String
    content: AWSJSON
    schedule: Schedule
    rrule: RRule
    recurrenceId: ID
    occurrenceId: String

    # Connections
    labels: Label @hasMany(relation: manyToMany)
    subtasks: Task @hasMany
    artifacts: Artifact @hasMany
    occurrences: Task @hasMany
  }

  type RRule {
    dtStart: AWSDateTime
    until: AWSDateTime
    ruleStr: String
  }

  type Schedule {
    startDate: DateTimeZone
    duration: String
    dueDate: DateTimeZone
  }

  type DateTimeZone {
    date: AWSDateTime
    timezone: String
  }

  type Label @model(operations: [upsert, delete]) {
    id: ID!
    name: String
    color: String
  }

  type File @model(operations: [upsert, delete]) {
    name: String
    size: Int
    url: AWSURL
    mimeType: String
  }

  type TimeTracker @model(operations: [upsert, delete]) {
    duration: String
    logs: [TimeLog!]
  }

  type TimeLog {
    action: TimeLogAction!
    timestamp: AWSTimestamp!
  }

  enum TimeLogAction {
    START
    PAUSE
    END
    LOG
  }

  union Artifact = File | TimeTracker

  type Viewer {
    user: User @hasOne(key: { ref: "identity.sub" })
    tasks: Task @hasMany(key: { ref: "identity.sub" }, index: "bySourceId")
    labels: Label @hasMany(key: { ref: "identity.sub" }, index: "bySourceId")
  }

  type Query {
    viewer: Viewer!
  }
`;

describe("GraphQLTransformer", () => {
  const transformer = createTransformer({
    definition: schema,
    mode: "development",
    outDir: resolve(dirname(fileURLToPath(import.meta.url)), "out"),
    dataSourceConfig: TEST_DS_CONFIG,
  });

  describe("createTransformer factory", () => {
    it("throws if empty definition", () => {
      expect(() => createTransformer({ definition: "" })).toThrow();
    });

    it("creates new GraphQLTransformer instance", () => {
      expect(transformer).toBeDefined();
      expect(transformer).toBeInstanceOf(GraphQLTransformer);
    });

    it("adds default plugins list", () => {
      expect(transformer.plugins).toHaveLength(6);
    });
  });

  describe("given invalid schema", () => {
    it("throws SchemaValidationError", () => {
      const transformer = createTransformer({
        definition: schema.replace("type Viewer", "type Viewer2"),
        dataSourceConfig: TEST_DS_CONFIG,
      });

      expect(() => transformer.transform()).toThrow(SchemaValidationError);
    });
  });

  describe("runs schema transformations", () => {
    transformer.transform();
    const context = transformer.context;

    it("copies extended fields", () => {
      const viewerType = context.document.getNode("Viewer") as ObjectNode;

      expect(viewerType.getField("user")).toBeInstanceOf(FieldNode);
      expect(viewerType.getField("tasks")).toBeInstanceOf(FieldNode);
    });

    describe("NodeInterfacePlugin trnsformations", () => {
      it("adds valid `node` field to Query", () => {
        const nodeField = (context.document.getNode("Query") as ObjectNode)?.getField("node");
        const nodeFieldTypename = (nodeField?.type as NamedTypeNode).name;

        expect(nodeField).toBeDefined();
        expect(nodeField).toBeInstanceOf(FieldNode);
        expect(nodeField?.hasArgument("id")).toStrictEqual(true);

        expect(nodeField?.type).toBeInstanceOf(NamedTypeNode);
        expect(nodeFieldTypename).toStrictEqual("Node");
      });

      it("extends models with Node interface", () => {
        const userNode = context.document.getNode("User") as ObjectNode;
        const taskNode = context.document.getNode("Task") as ObjectNode;

        expect(userNode.hasInterface("Node")).toBeTruthy();
        expect(taskNode.hasInterface("Node")).toBeTruthy();
      });
    });

    describe("ModelPlugin", () => {
      it("created query fields for model", () => {
        const queryNode = context.document.getQueryNode();

        expect(queryNode.getField("getUser")).toBeInstanceOf(FieldNode);
        expect(queryNode.getField("listUsers")).toBeInstanceOf(FieldNode);

        expect(queryNode.getField("getTask")).not.toBeDefined();
        expect(queryNode.getField("listTasks")).not.toBeDefined();
      });
      it("created mutation fields for model", () => {
        const mutationNode = context.document.getMutationNode();

        expect(mutationNode.getField("createUser")).toBeInstanceOf(FieldNode);
        expect(mutationNode.getField("updateUser")).toBeInstanceOf(FieldNode);
        expect(mutationNode.getField("deleteUser")).toBeInstanceOf(FieldNode);

        expect(mutationNode.getField("upsertTask")).toBeInstanceOf(FieldNode);
        expect(mutationNode.getField("deleteTask")).toBeInstanceOf(FieldNode);
      });

      it("created operation inputs & types", () => {
        expect(context.document.getNode("CreateUserInput")).toBeDefined();
        expect(context.document.getNode("UpdateUserInput")).toBeDefined();
        expect(context.document.getNode("DeleteUserInput")).toBeDefined();
        expect(context.document.getNode("UpsertTaskInput")).toBeDefined();
        expect(context.document.getNode("DeleteTaskInput")).toBeDefined();
      });
    });
  });
});
