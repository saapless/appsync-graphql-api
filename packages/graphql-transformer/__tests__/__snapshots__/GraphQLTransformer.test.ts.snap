// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQLTransformer runs schema transformations return result 1`] = `
{
  "schema": "enum UserStatus {
  ACTIVE
  DISABLED
  SUSPENDED
}

type User implements Node {
  id: ID!
  firstName: String
  lastName: String
  email: AWSEmail
  picture: AWSURL
  status: UserStatus
  createdAt: String
  updatedAt: String
}

type Task implements Node {
  id: ID!
  title: String
  content: AWSJSON
  schedule: Schedule
  rrule: RRule
  recurrenceId: ID
  occurrenceId: String
  labels(filter: LabelEdgeFilterInput, first: Int, after: String, sort: SortDirection): LabelConnection!
  occurrences(filter: TaskFilterInput, first: Int, after: String, sort: SortDirection): TaskConnection!
  subtasks(filter: TaskFilterInput, first: Int, after: String, sort: SortDirection): TaskConnection!
  artifacts(filter: ArtifactEdgeFilterInput, first: Int, after: String, sort: SortDirection): ArtifactConnection!
  createdAt: String
  updatedAt: String
}

type RRule {
  dtStart: AWSDateTime
  until: AWSDateTime
  ruleStr: String
}

type Schedule {
  startDate: DateTimeZone
  duration: String
  dueDate: DateTimeZone
}

type DateTimeZone {
  date: AWSDateTime
  timezone: String
}

type Label implements Node {
  id: ID!
  name: String
  color: String
  createdAt: String
  updatedAt: String
}

type File implements Node {
  name: String
  size: Int
  url: AWSURL
  mimeType: String
  id: ID!
  createdAt: String
  updatedAt: String
}

type TimeTracker implements Node {
  duration: String
  logs: [TimeLog!]
  id: ID!
  createdAt: String
  updatedAt: String
}

type TimeLog {
  action: TimeLogAction!
  timestamp: AWSTimestamp!
}

enum TimeLogAction {
  START
  PAUSE
  END
  LOG
}

union Artifact = File | TimeTracker

type Viewer {
  user: User
  tasks(filter: TaskFilterInput, first: Int, after: String, sort: SortDirection): TaskConnection!
  labels(filter: LabelFilterInput, first: Int, after: String, sort: SortDirection): LabelConnection!
}

type Query {
  viewer: Viewer!
  node(id: ID!): Node
  getUser(id: ID!): User
  listUsers(filter: UserFilterInput, first: Int, after: String, sort: SortDirection): UserConnection!
}

interface Node {
  id: ID!
}

type PageInfo {
  hasNextPage: Boolean
  hasPreviousPage: Boolean
  startCursor: String
  endCursor: String
}

input SizeFilterInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int!]
}

input StringFilterInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  in: [String!]
  contains: String
  notContains: String
  between: [String!]
  beginsWith: String
  attributeExists: Boolean
  size: SizeFilterInput
}

input IntFilterInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  in: [Int!]
  between: [Int!]
  attributeExists: Boolean
}

input FloatFilterInput {
  ne: Float
  eq: Float
  le: Float
  lt: Float
  ge: Float
  gt: Float
  in: [Float!]
  between: [Float!]
  attributeExists: Boolean
}

input BooleanFilterInput {
  ne: Boolean
  eq: Boolean
  attributeExists: Boolean
}

input IDFilterInput {
  ne: ID
  eq: ID
  in: [ID!]
  attributeExists: Boolean
}

input ListFilterInput {
  contains: String
  notContains: String
  size: SizeFilterInput
}

enum SortDirection {
  ASC
  DESC
}

type Mutation {
  createUser(input: CreateUserInput!): User
  updateUser(input: UpdateUserInput!): User
  deleteUser(id: ID!): User
  upsertTask(input: UpsertTaskInput!): Task
  deleteTask(id: ID!): Task
  createLabelEdge(input: LabelEdgeInput!): LabelEdge
  deleteLabelEdge(input: LabelEdgeInput!): LabelEdge
  createArtifactEdge(input: ArtifactEdgeInput!): ArtifactEdge
  deleteArtifactEdge(input: ArtifactEdgeInput!): ArtifactEdge
  upsertLabel(input: UpsertLabelInput!): Label
  deleteLabel(id: ID!): Label
  upsertFile(input: UpsertFileInput!): File
  deleteFile(id: ID!): File
  upsertTimeTracker(input: UpsertTimeTrackerInput!): TimeTracker
  deleteTimeTracker(id: ID!): TimeTracker
}

input CreateUserInput {
  id: ID!
  firstName: String
  lastName: String
  email: AWSEmail
  picture: AWSURL
  createdAt: String
  updatedAt: String
}

input UpdateUserInput {
  id: ID!
  firstName: String
  lastName: String
  email: AWSEmail
  picture: AWSURL
  createdAt: String
  updatedAt: String
}

input DateTimeZoneInput {
  date: AWSDateTime
  timezone: String
}

input ScheduleInput {
  startDate: DateTimeZoneInput
  duration: String
  dueDate: DateTimeZoneInput
}

input RrUleInput {
  dtStart: AWSDateTime
  until: AWSDateTime
  ruleStr: String
}

input UpsertTaskInput {
  id: ID!
  title: String
  content: AWSJSON
  schedule: ScheduleInput
  rrule: RrUleInput
  recurrenceId: ID
  occurrenceId: String
  sourceId: ID
  createdAt: String
  updatedAt: String
}

type LabelConnection {
  edges: [LabelEdge!]!
  pageInfo: PageInfo!
}

type LabelEdge {
  cursor: String
  node: Label
}

input LabelEdgeFilterInput {
  createdAt: StringFilterInput
  updatedAt: StringFilterInput
  and: [LabelEdgeFilterInput]
  or: [LabelEdgeFilterInput]
  not: LabelEdgeFilterInput
}

input LabelEdgeInput {
  sourceId: ID!
  targetId: ID!
}

type TaskConnection {
  edges: [TaskEdge!]!
  pageInfo: PageInfo!
}

type TaskEdge {
  cursor: String
  node: Task
}

input TaskFilterInput {
  id: IDFilterInput
  title: StringFilterInput
  content: StringFilterInput
  recurrenceId: IDFilterInput
  occurrenceId: StringFilterInput
  createdAt: StringFilterInput
  updatedAt: StringFilterInput
  and: [TaskFilterInput]
  or: [TaskFilterInput]
  not: TaskFilterInput
}

type ArtifactConnection {
  edges: [ArtifactEdge!]!
  pageInfo: PageInfo!
}

type ArtifactEdge {
  cursor: String
  node: Artifact
}

input ArtifactEdgeFilterInput {
  createdAt: StringFilterInput
  updatedAt: StringFilterInput
  and: [ArtifactEdgeFilterInput]
  or: [ArtifactEdgeFilterInput]
  not: ArtifactEdgeFilterInput
}

input ArtifactEdgeInput {
  sourceId: ID!
  targetId: ID!
}

input UpsertLabelInput {
  id: ID!
  name: String
  color: String
  sourceId: ID
  createdAt: String
  updatedAt: String
}

input UpsertFileInput {
  name: String
  size: Int
  url: AWSURL
  mimeType: String
  sourceId: ID
  id: ID!
  createdAt: String
  updatedAt: String
}

input TimeLogInput {
  action: TimeLogAction
  timestamp: AWSTimestamp
}

input UpsertTimeTrackerInput {
  duration: String
  logs: TimeLogInput
  sourceId: ID
  id: ID!
  createdAt: String
  updatedAt: String
}

input LabelFilterInput {
  id: IDFilterInput
  name: StringFilterInput
  color: StringFilterInput
  createdAt: StringFilterInput
  updatedAt: StringFilterInput
  and: [LabelFilterInput]
  or: [LabelFilterInput]
  not: LabelFilterInput
}

type UserConnection {
  edges: [UserEdge!]!
  pageInfo: PageInfo!
}

type UserEdge {
  cursor: String
  node: User
}

input UserStatusFilterInput {
  eq: UserStatus
  ne: UserStatus
  in: [UserStatus!]
  attributeExists: Boolean
}

input UserFilterInput {
  id: IDFilterInput
  firstName: StringFilterInput
  lastName: StringFilterInput
  email: StringFilterInput
  picture: StringFilterInput
  status: UserStatusFilterInput
  createdAt: StringFilterInput
  updatedAt: StringFilterInput
  and: [UserFilterInput]
  or: [UserFilterInput]
  not: UserFilterInput
}",
}
`;
